File: ObjcFrame/ObjcFrame/Base/View/BWSearchBarView(搜索框)/BWSearchBarView.m
//
//  BWSearchBarView.m
//  MengTianXia
//
//  Created by zl on 2019/8/14.
//  Copyright © 2019 qttx. All rights reserved.
//

#import "BWSearchBarView.h"

@interface BWSearchBarView () <UITextFieldDelegate>
@property (strong, nonatomic) UIView *searchView;
@property (strong, nonatomic) UIImageView *searchImage;
@end

@implementation BWSearchBarView

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        
        self.searchView = [[UIView alloc] init];
        self.searchView.backgroundColor = BWColor(244, 244, 244);
        
        self.searchImage = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"search_bar_icon"]];
        self.searchImage.contentMode = UIViewContentModeScaleAspectFit;
        
        self.searchField = [[UITextField alloc] init];
        self.searchField.font = [UIFont systemFontOfSize:14];
        self.searchField.textColor = TEXT_BLACK_COLOR;
        self.searchField.tintColor = TEXT_GRAY_COLOR;
        self.searchField.clearButtonMode = UITextFieldViewModeWhileEditing;
        self.searchField.returnKeyType = UIReturnKeySearch;
        self.searchField.delegate = self;
        [self.searchField addTarget:self action:@selector(searchTextFieldDidChange:) forControlEvents:UIControlEventEditingChanged];
        
        [self.searchView addSubview:self.searchImage];
        [self.searchView addSubview:self.searchField];
        [self addSubview:self.searchView];
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    self.searchView.frame = CGRectMake(0, 0, self.width, self.height);
    self.searchView.layer.cornerRadius = self.height * 0.5;
    self.searchView.layer.masksToBounds = YES;
    
    self.searchImage.frame = CGRectMake(10, 0, 15, 15);
    self.searchField.frame = CGRectMake(CGRectGetMaxX(self.searchImage.frame) + 10, 0, self.width - self.searchImage.width - 20, self.searchView.height);
    self.searchImage.centerY = self.searchField.centerY;
}

- (void)setPlaceholderString:(NSString *)placeholderString {
    self.searchField.placeholder = placeholderString;
}

- (void)setSearchString:(NSString *)searchString {
    self.searchField.text = searchString;
}


#pragma mark - UITextFieldDelegate

/// 开始搜索
- (void)textFieldDidBeginEditing:(UITextField *)textField {
    if ([self.delegate respondsToSelector:@selector(searchBarView:beginSearchWithSearchTextField:)]) {
        [self.delegate searchBarView:self beginSearchWithSearchTextField:textField];
    }
}

/// 结束搜索
- (void)textFieldDidEndEditing:(UITextField *)textField {
    if ([self.delegate respondsToSelector:@selector(searchBarView:endSearchWithSearchTextField:)]) {
        [self.delegate searchBarView:self endSearchWithSearchTextField:textField];
    }
}

/// 搜索关键字改变
- (void)searchTextFieldDidChange:(UITextField *)textField {
    if ([self.delegate respondsToSelector:@selector(searchBarView:searchTextFieldDidChange:)]) {
        [self.delegate searchBarView:self searchTextFieldDidChange:textField.text];
    }
}

/// 点击搜索按钮
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if ([self.delegate respondsToSelector:@selector(searchBarView:didClickReturnKeyboardWithKeyword:)]) {
        [self.delegate searchBarView:self didClickReturnKeyboardWithKeyword:textField.text];
        [textField resignFirstResponder];
    }
    return YES;
}

@end

File: SwiftFrame/MengTianXia/Classes/Home(首页)/View/BWSearchBarView.swift
//
//  BWSearchBarView.swift
//  MengTianXia
//
//  Created by zl on 2019/7/26.
//  Copyright © 2019 LuMeng. All rights reserved.
//

import UIKit

class BWSearchBarView: UIView {
    var searchView: UIView
    var searchImage: UIImageView
    var searchField: UITextField
    
    override init(frame: CGRect) {
        searchView = UIView()
        searchView.backgroundColor = BWColor(245, 245, 245)
        searchView.layer.cornerRadius = 18
        searchView.layer.masksToBounds = true
        
        searchImage = UIImageView(image: UIImage(named: "search"))
        searchImage.contentMode = .scaleAspectFit
        
        searchField = UITextField()
        searchField.font = UIFont.systemFont(ofSize: 14)
        searchField.textColor = BlackTextColor
        searchField.placeholder = "搜索你需要的房源"
        searchField.clearButtonMode = .whileEditing
        
        searchView.addSubview(searchImage)
        searchView.addSubview(searchField)
        
        super.init(frame: frame)
        
        addSubview(searchView)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        let searchH: CGFloat = 36.0
        
        searchView.frame = CGRect(x: 0, y: (frame.height - 36) * 0.5, width: frame.width, height: searchH)
        searchImage.frame = CGRect(x: 10, y: 0, width: searchH * 0.5, height: searchH * 0.5)
        searchField.frame = CGRect(x: searchImage.frame.width + 16, y: 0, width: frame.width - searchImage.frame.width - 26, height: searchView.frame.height)
        searchImage.center.y = searchField.center.y
    }
    
}
